[Fleet Monitor] petry-projects/.github-private — .github/workflows/feature-ideation.yml

## 🟠 DEGRADED: `petry-projects/.github-private` / `.github/workflows/feature-ideation.yml`

| Metric | Value |
|---|---|
| **Failure Rate** | 33.3% (2 / 6 runs) |
| **Status** | DEGRADED |
| **p50 Duration** | 95s |
| **p95 Duration** | 502s |
| **Successful Runs** | 4 |
| **Cancelled Runs** | 0 |

**Workflow:** [`.github/workflows/feature-ideation.yml`](https://github.com/petry-projects/.github-private/actions)

> _Last updated by [Fleet Monitor run](https://github.com/petry-projects/.github-private/actions/runs/28317708645) on 2026-06-28._
> _Threshold: failure rate > 10% over the monitored window._

## Dev-Lead Implementation Plan

**Issue:** #985 — [Fleet Monitor] petry-projects/.github-private — .github/workflows/feature-ideation.yml

### Investigation — root cause of the DEGRADED (33.3%) signal

Pulling the run history for `feature-ideation.yml`, the failures cluster almost entirely on the **`discussion`** trigger:

| Failure | Event | Root cause |
|---|---|---|
| 06-21 → 06-27 14:52 (all discussion runs) | `discussion` | `##[error]Action failed with error: Unsupported event type: discussion` — claude-code-action aborts on a `discussion` event context when the `ideate` job invokes the reusable inline. |
| 06-26 (×2) | `discussion`, `schedule` | `startup_failure` — transient reusable-tag/infra resolution (`feature-ideation/next`). |
| 06-12 (×1) | `schedule` | transient `claude-code-action` failure in the LLM research job. |

**The dominant, recurring failure (`Unsupported event type: discussion`) is already fixed** by the #963 redispatch bridge that is present in `main` today: a `redispatch` job handles the `discussion` event by re-invoking the stub via `workflow_dispatch`, and the reusable-calling `ideate` job is gated `if: github.event_name != 'discussion'`. The first post-fix discussion run (06-27 15:54, sha `6fceab6`) succeeded (redispatch ✓, ideate skipped). The 2 failures in the monitor's window are **pre-fix** and will age out.

The remaining sporadic failures (`startup_failure`, the single 06-12 schedule failure) are **transient** (upstream reusable-tag/infra + inherent LLM flakiness) and are **not structurally fixable within this thin-caller stub** — its header forbids changing the `uses:` line, triggers, permissions, or secrets; retry/timeout logic lives in the reusable in `petry-projects/.github`.

### Scope
The workflow fix is already in place and CI-enforced (`tests/test_feature_ideation.bats` runs via `lint.yml`). The one genuine gap: the existing guard test only greps for `!= 'discussion'` **anywhere** in the file, so a regression that moved the `uses:` reusable call to an **unguarded** job could still pass. Close that gap with a **structural** (yq) assertion binding the off-discussion gate to the actual reusable-calling job — hardening against silent recurrence of the exact regression that produced this tracker. Test-only; no production behavior change; respects the thin-caller contract.

### Implementation Checklist
- [ ] `tests/test_feature_ideation.bats` — add yq-based structural test(s) asserting the job that `uses:` the `feature-ideation-reusable.yml` reusable is gated `if: github.event_name != 'discussion'`, and that the `redispatch` job (the discussion handler) does not itself call the reusable.

### Tests to Write
- [ ] "the reusable-calling job is structurally gated off the discussion event" (yq)
- [ ] "the discussion-event handler job does not invoke the reusable inline" (yq)
- [ ] Demonstrate red phase against a mutated copy before locking green.

### Test Command
`bash scripts/run-bats.sh tests/test_feature_ideation.bats` (full guard suite), then `bash .dev-lead/scripts/dev-lead-lint.sh`

## Dev-Lead: Implementation Complete

### Root-cause summary
The DEGRADED signal was driven almost entirely by the **`discussion`** trigger: every discussion-triggered run failed with `##[error]Action failed with error: Unsupported event type: discussion` (claude-code-action aborts on a `discussion` event context). **That root cause is already fixed in `main`** by the #963 redispatch bridge (a `redispatch` job re-invokes the stub via `workflow_dispatch`; the reusable-calling `ideate` job is gated `if: github.event_name != 'discussion'`). The first post-fix discussion run (06-27 15:54, sha `6fceab6`) succeeded (redispatch ✓, ideate skipped). The 2 failures in the monitor window are **pre-fix** and age out on their own.

The other failures are transient and **not** structurally fixable inside this thin-caller stub (whose header forbids changing `uses:`/triggers/permissions/secrets): two `startup_failure`s (06-26) from reusable-tag/infra resolution, and one 06-12 `schedule` failure from inherent `claude-code-action` LLM flakiness.

### What changed and why
No production workflow change was warranted — the fix is in place. The one real gap: the existing guard (`tests/test_feature_ideation.bats`) only `grep`s for `!= 'discussion'` **anywhere** in the file, so a regression that moved the `uses:` reusable call onto an unguarded (or discussion-handling) job would slip through while resurrecting the exact abort. I added two **structural** (yq) assertions that bind the gate to the actual reusable-calling job. This guard is already CI-enforced (`lint.yml` runs this file).

This was validated TDD-style: the new assertion goes **red** on a mutated copy whose `ideate` guard is removed (while the old grep test gives a false-negative **pass**), and **green** on the current file.

### Plan Execution
- [x] Investigated run history; identified `Unsupported event type: discussion` as the dominant failure and confirmed it's already fixed + CI-guarded
- [x] Confirmed remaining failures are transient/upstream and out of scope for the thin stub
- [x] Added yq-based structural regression tests (red→green proven)
- [x] Full bats guard suite + repo lint pass

### Test Results
```
$ bash scripts/run-bats.sh tests/test_feature_ideation.bats
1..14
ok 1 feature-ideation.yml exists
...
ok 11 feature-ideation.yml ideate job no longer runs on the discussion event
ok 12 feature-ideation.yml sources target_discussion from inputs.target_discussion
ok 13 feature-ideation.yml: every reusable-calling job is gated off the discussion event
ok 14 feature-ideation.yml: the discussion handler reaches the reusable only via the dispatch bridge

$ bash .dev-lead/scripts/dev-lead-lint.sh
  [lint] shellcheck on 82 file(s)...
  [lint] validating 4 agent profile(s)...
  [lint] validating .github/CODEOWNERS...
  [lint] all checks passed

$ bash scripts/run-bats.sh tests/    # full suite
ok: 1580   not ok: 2
```
The 2 full-suite failures are in unrelated files I did not touch (`test_review_cycle.bats` MCP-debug handshake, `test_batch_fallback.bats` Gemini-fallback) and are environment-dependent (engine availability/mocking in this sandbox) — not caused by this change (diff is isolated to `tests/test_feature_ideation.bats`).

### Files Changed
- `tests/test_feature_ideation.bats`: +2 yq-based structural tests pinning that every reusable-calling job is gated off the `discussion` event and that the discussion handler does not invoke the reusable inline.

### Notes
- Test-only change; no production behavior change; respects the thin-caller stub contract.
- The workflow's failure rate recovers mechanically as the pre-fix discussion failures leave the monitoring window; this PR prevents the same regression from silently returning.
- Possible follow-up (out of scope here): a live-trigger canary for `feature-ideation` analogous to `initiative-planner-canary.yml` (#822), which guards the same `Unsupported event type: discussion` class for the initiative-planner.

--- a/tests/test_feature_ideation.bats
+++ b/tests/test_feature_ideation.bats
@@ -112,10 +112,12 @@ setup() {
 @test "feature-ideation.yml: every reusable-calling job is gated off the discussion event" {
   # At least one job must call the reusable (guard against a vacuous pass)…
   local total ungated
-  total="$(yq '[.jobs[] | select(.uses | test("feature-ideation-reusable.yml"))] | length' "$FEATURE_IDEATION_YML")"
+  total="$(yq '[.jobs[] | select((.uses // "") | test("feature-ideation-reusable.yml"))] | length' "$FEATURE_IDEATION_YML")"
   [ "$total" -ge 1 ]
   # …and none of them may run on a discussion event (claude-code-action aborts there).
-  ungated="$(yq '[.jobs[] | select(.uses | test("feature-ideation-reusable.yml")) | select((.if // "") | test("event_name *!= *.discussion.") | not)] | length' "$FEATURE_IDEATION_YML")"
+  # A job is "ungated" if its if: lacks the event_name != 'discussion' guard, OR if it
+  # contains || (which could let a second OR-branch re-enable execution on discussions).
+  ungated="$(yq '[.jobs[] | select((.uses // "") | test("feature-ideation-reusable.yml")) | select(((.if // "") | test("event_name *!= *.discussion.") | not) or ((.if // "") | test("[|][|]")))] | length' "$FEATURE_IDEATION_YML")"
   [ "$ungated" -eq 0 ]
 }
 

--- a/tests/test_feature_ideation.bats
+++ b/tests/test_feature_ideation.bats
@@ -98,3 +98,30 @@ setup() {
 @test "feature-ideation.yml sources target_discussion from inputs.target_discussion" {
   grep -qE 'target_discussion:[[:space:]]*"?\$\{\{[[:space:]]*inputs\.target_discussion' "$FEATURE_IDEATION_YML"
 }
+
+# ── #985: structural guard binding the off-discussion gate to the reusable call ──
+# The grep-anywhere check above passes as long as the "!= 'discussion'" string
+# exists *somewhere* in the file — it does NOT verify the gate sits on the job
+# that actually invokes the reusable. A regression that moved the `uses:` call
+# onto an unguarded (or discussion-handling) job would slip through it while
+# resurrecting the exact "Unsupported event type: discussion" failure that
+# degraded this workflow (#985). These yq-based tests pin the invariant
+# structurally: every job that calls the reusable must be gated off the
+# discussion event, and the discussion handler must not call it inline.
+
+@test "feature-ideation.yml: every reusable-calling job is gated off the discussion event" {
+  # At least one job must call the reusable (guard against a vacuous pass)…
+  local total ungated
+  total="$(yq '[.jobs[] | select(.uses | test("feature-ideation-reusable.yml"))] | length' "$FEATURE_IDEATION_YML")"
+  [ "$total" -ge 1 ]
+  # …and none of them may run on a discussion event (claude-code-action aborts there).
+  ungated="$(yq '[.jobs[] | select(.uses | test("feature-ideation-reusable.yml")) | select((.if // "") | test("event_name *!= *.discussion.") | not)] | length' "$FEATURE_IDEATION_YML")"
+  [ "$ungated" -eq 0 ]
+}
+
+@test "feature-ideation.yml: the discussion handler reaches the reusable only via the dispatch bridge" {
+  # The redispatch job handles the discussion event; it must NOT call the
+  # reusable inline (which would re-introduce the abort) — it re-dispatches
+  # under workflow_dispatch instead.
+  [ "$(yq '.jobs.redispatch.uses' "$FEATURE_IDEATION_YML")" = "null" ]
+}

